### PR TITLE
[generate] Fix encoder decoder models attention mask

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

https://github.com/huggingface/transformers/pull/35852 introduced a bug where the attention mask is lost for encoder-decoder models.
This PR implements a naïve fix.

cf https://github.com/huggingface/transformers/issues/36020

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for noticing it's broken, should've triggered some slow tests before merging!

I agree that we'd need to pass both attention masks for encoder-decoder models if available. And the attention mask in `prepare_inputs` should come from decoder for further manipulations no top

Overall LGTM but would be cleaner if we could pass `encoder-mask` from within `kwargs` in step 7, imo. CC @gante for one more opinion on generation

**Comment by eustlb** [comment]
agree that having a `encoder_mask` would be cleaner, but it diverges a bit with the current convention in the codebase where encoder-decoder model forward's use `attention_mask` for the encoder and ` decoder_attention_mask` otherwise. It is also how it is currently documents:
https://github.com/huggingface/transformers/blob/9d2056f12b66e64978f78a2dcb023f65b2be2108/src/transformers/generation/utils.py#L1957-L1959

**Review by gante — APPROVED** [review]
LGTM, thank you for the fix 🙏 

P.S.: can we somehow add a fast test to ensure this doesn't happen again? This is a pretty basic feature, and should be covered by fast tests

**Comment by gante** [comment]
@zucchini-nlp As @eustlb wrote, historically the naming convention is for the `attention_mask` to attend to the input (back when there were no multimodal models) 👀

**Comment by gante** [comment]
@eustlb I'm going to rerun the tests until CI gets green, and then merge this PR -- not having this fix is blocking me :D
